dds/DCPS/RTPS/Spdp.cpp

```diff
@@ -965,16 +1025,34 @@ Spdp::data_received(const DataSubmessage& data,
     return;
   }
 
-  handle_participant_data(
-    (data.inlineQos.length() && disposed(data.inlineQos)) ?
-      DCPS::DISPOSE_INSTANCE : DCPS::SAMPLE_DATA,
-    pdata, to_opendds_seqnum(data.writerSN), from, false);
-
+  const bool from_relay = from == config_->spdp_rtps_relay_address();
+  if (DCPS::DCPS_debug_level) {
+    ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) Spdp::data_received -  from: %C\n"), DCPS::LogAddr(from).c_str()));
+    ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) Spdp::data_received - relay: %C\n"), DCPS::LogAddr(config_->spdp_rtps_relay_address()).c_str()));
+  }
 #ifdef OPENDDS_SECURITY
+  if (!from_relay && !ip_in_locator_list(from, pdata.participantProxy.metatrafficUnicastLocatorList) && !ip_in_AgentInfo(from, plist)) {
+    if (DCPS::DCPS_debug_level) {
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) Spdp::data_received - dropped IP: %C\n"), DCPS::LogAddr(from).c_str()));
+    }
+    return;
+  }
   if (!is_security_enabled()) {
     process_participant_ice(plist, pdata, guid);
   }
+#elif !defined OPENDDS_SAFETY_PROFILE
+  if (!from_relay && !ip_in_locator_list(from, pdata.participantProxy.metatrafficUnicastLocatorList)) {
+    if (DCPS::DCPS_debug_level) {
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) Spdp::data_received - IP not in locator list: %C\n"), DCPS::LogAddr(from).c_str()));
+    }
+    return;
+  }
 #endif
+
+  handle_participant_data(
```

Comment:
Did you mean to move the call to handle_participant_data() to after process_participant_ice()? That change in order doesn't seem related to checking the spdp received addresses against locators.

Reply:
Justin added this:

Acceptance Criteria

Implement the metatraffic locator check described in the parent issue (378 - see “possible solutions #1”) with appropriate logging.

Here is the sketch for ICE.

We may be tempted to move this block 

#ifdef OPENDDS_SECURITY
  if (!is_security_enabled()) {
    process_participant_ice(plist, pdata, guid);
  }
#endif

before the call to handle_participant_data.  However, we can (and should) assume that we get an SPDP message through the relay or some other legitimate means first.

2.  The definition of process_participant_ice refers to members named sedp_info_ and spdp_info_.  These members are an ICE::AgentInfo (dds/DCPS/Ice.h).  This contains a list of candidates and each candidate is an address.  Extend the  ip_in_locator_list to check these candidates.